velox/tool/trace/PartitionedOutputReplayer.cpp

```diff
@@ -0,0 +1,172 @@
+  VELOX_CHECK(bufferManager->getData(
+      taskId,
+      destination,
+      exec::PartitionedOutput::kMinDestinationSize,
+      sequence,
+      [result = std::make_shared<
+           folly::Promise<std::vector<std::unique_ptr<folly::IOBuf>>>>(
+           std::move(promise))](
+          std::vector<std::unique_ptr<folly::IOBuf>> pages,
+          int64_t /*inSequence*/,
+          std::vector<int64_t> /*remainingBytes*/) {
+        result->setValue(std::move(pages));
+      }));
+  auto future = std::move(semiFuture).via(executor);
+  future.wait(std::chrono::seconds{10});
+  VELOX_CHECK(future.isReady());
+  return std::move(future).value();
+}
+} // namespace
+
+void consumeAllData(
+    const std::shared_ptr<exec::OutputBufferManager>& bufferManager,
+    const std::string& taskId,
+    uint32_t numPartitions,
+    folly::Executor* executor,
+    std::function<void(uint32_t, std::unique_ptr<folly::IOBuf>)> consumer) {
+  std::vector<std::thread> consumerThreads;
+  std::vector<int64_t> sequences;
+  consumerThreads.reserve(numPartitions);
+  sequences.reserve(numPartitions);
+  sequences.resize(numPartitions);
+  for (uint32_t i = 0; i < numPartitions; i++) {
+    consumerThreads.push_back(std::thread([&, partition = i]() {
+      bool finished{false};
+      while (!finished) {
+        std::vector<std::unique_ptr<folly::IOBuf>> pages;
+        {
+          pages = getData(
+              bufferManager, taskId, partition, sequences[partition], executor);
+        }
+        for (auto& page : pages) {
+          if (page) {
+            consumer(partition, std::move(page));
+            sequences[partition]++;
+          } else {
+            // Null page indicates this buffer is finished.
+            bufferManager->deleteResults(taskId, partition);
+            finished = true;
+          }
+        }
+      }
+    }));
+  }
+
+  for (auto& thread : consumerThreads) {
+    thread.join();
+  }
+}
+
+PartitionedOutputReplayer::PartitionedOutputReplayer(
+    const std::string& rootDir,
+    const std::string& taskId,
+    const std::string& nodeId,
+    const int32_t pipelineId,
+    const std::string& operatorType,
+    std::function<void(uint32_t, std::unique_ptr<folly::IOBuf>)> consumer)
+    : OperatorReplayerBase(rootDir, taskId, nodeId, pipelineId, operatorType),
+      originalNode_(dynamic_cast<const core::PartitionedOutputNode*>(
+          core::PlanNode::findFirstNode(
+              planFragment_.get(),
+              [this](const core::PlanNode* node) {
+                return node->id() == nodeId_;
+              }))),
+      consumer_(consumer) {
+  VELOX_CHECK_NOT_NULL(originalNode_);
+}
+
+RowVectorPtr PartitionedOutputReplayer::run() {
+  auto task = Task::create(
+      "local://partitioned-output-replayer",
+      core::PlanFragment{createPlan()},
+      0,
+      createQueryContext(queryConfigs_, executor_.get()),
+      Task::ExecutionMode::kParallel);
+  task->start(maxDrivers_);
+
+  consumeAllData(
```

Comment:
Could we just pass task id to consumeAllData for now as it is a class  method?

Reply:
Intended to put it as helper method for use in unit test. Otherwise we need to duplicate it.